File: app/src/main/java/com/billpayments/features/saved_billers/data/SavedBillersRepository.kt
package com.billpayments.features.saved_billers.data

import com.billpayments.core.storage.KeyValueStore
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

interface SavedBillersRepository {
    fun restore(userId: String): SavedBillers
    suspend fun persist(userId: String, value: SavedBillers)
}

@Serializable
private data class SavedBillerDto(val billerId: String, val account: String, val nickname: String)

class StoredSavedBillersRepository(private val store: KeyValueStore) : SavedBillersRepository {

    override fun restore(userId: String): SavedBillers {
        val raw = store.read(keyFor(userId)) ?: return SavedBillers.EMPTY
        val dtos = Json.decodeFromString<List<SavedBillerDto>>(raw)
        return SavedBillers(dtos.map { SavedBiller(it.billerId, it.account, it.nickname) })
    }

    override suspend fun persist(userId: String, value: SavedBillers) {
        val dtos = value.items.map { SavedBillerDto(it.billerId, it.account, it.nickname) }
        store.write(keyFor(userId), Json.encodeToString(dtos))
    }

    private fun keyFor(userId: String) = "savedBillers.$userId"
}

File: BillPayments/Features/SavedBillers/Data/SavedBillersRepository.swift
import Foundation

protocol SavedBillersRepository {
    func restore(userId: String) -> SavedBillers
    func persist(userId: String, value: SavedBillers) async throws
}

final class StoredSavedBillersRepository: SavedBillersRepository {
    private let store: KeyValueStore

    init(store: KeyValueStore) {
        self.store = store
    }

    func restore(userId: String) -> SavedBillers {
        guard let raw = store.read(keyFor(userId)),
              let data = raw.data(using: .utf8),
              let items = try? JSONDecoder().decode([SavedBiller].self, from: data)
        else { return .empty }
        return SavedBillers(items: items)
    }

    func persist(userId: String, value: SavedBillers) async throws {
        let data = try JSONEncoder().encode(value.items)
        guard let raw = String(data: data, encoding: .utf8) else { throw StorageError() }
        try await store.write(keyFor(userId), raw)
    }

    private func keyFor(_ userId: String) -> String { "savedBillers.\(userId)" }
}
